[AMDGPU] Eliminate likely-spurious execz checks

ritter-x2a opened the pull request:
Currently, we introduce branches to skip conditionally executed instructions if the EXEC mask is zero and only eliminate them if the scheduling model says that executing the skipped instructions is cheaper than taking the branch instruction.

This patch adds branch weights to branches where the threads of a wavefront are likely (according to a heuristic) to have dynamically varying values for the branch condition. This causes SIPreEmitPeephole to eliminate the corresponding execz branch.

Currently, this is implemented as a new middle end pass with a rather optimistic heuristic, to gather initial feedback.

ritter-x2a committed:
[AMDGPU] Eliminate likely-spurious execz checks

Currently, we introduce branches to skip conditionally executed
instructions if the EXEC mask is zero and only eliminate them if the
scheduling model says that executing the skipped instructions is cheaper
than taking the branch instruction.

This patch adds branch weights to branches where the threads of a
wavefront are likely (according to a heuristic) to have dynamically
varying values for the branch condition. This causes SIPreEmitPeephole
to eliminate the corresponding execz branch.

Currently, this is implemented as a new middle end pass with a rather
optimistic heuristic, to gather initial feedback.

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUAnnotateVaryingBranchWeights.cpp:
```diff
@@ -0,0 +1,201 @@
+      // intrinsics connect it via data flow with the actual condition
+      // (even for the amdgcn_else intrinsic, via the matching amdgcn_if
+      // intrinsic), so isLikelyVarying still produces meaningful results.
+      if (isLikelyVarying(cast<BranchInst>(Term)->getCondition()))
```
I don't understand the big picture here. `isLikelyVarying` just seems to be reimplementing divergence analysis, by checking for things that either _are_ or _use_ sources of divergence. But don't we already know that the branch is divergent, because SIAnnotateControl only inserts intrinsics like amdgcn_if in divergent control flow?

ritter-x2a replied:
The goal is to use a heuristic there to determine when branches are likely to be varying, for instance when they are based on (the lower bits of) the thread id. The current implementation here is very likely not the heuristic we want. At this point, the point of this PR is to gather initial feedback on where and how I use the heuristic before tuning it.

ritter-x2a replied:
Well, technically not quite, right? Uniformity analysis does more than just follow data dependences. Via control dependences (e.g. as a phi at the control flow join of a non-uniform branch), a Value could be non-uniform but this heuristic wouldn't trigger (but we'd probably want to be more restrictive than that as well).

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUPassRegistry.def:
```diff
@@ -65,6 +65,9 @@ FUNCTION_PASS("amdgpu-unify-divergent-exit-nodes",
               AMDGPUUnifyDivergentExitNodesPass())
 FUNCTION_PASS("amdgpu-usenative", AMDGPUUseNativeCallsPass())
 FUNCTION_PASS("si-annotate-control-flow", SIAnnotateControlFlowPass(*static_cast<const GCNTargetMachine *>(this)))
+FUNCTION_PASS("amdgpu-annotate-varying-branch-weights",
```
Passes should be sorted alphabetically here

ritter-x2a replied:
Done in 345269b1523b1c2af48dc3798519031bf30d218d.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAnnotateVaryingBranchWeights.cpp:
```diff
@@ -0,0 +1,201 @@
+    // Look for conditional branches whose condition is an ExtractValueInst
+    // that extracts the return value of a call to the amdgcn_if or amdgcn_else
+    // intrinsic.
+    if (match(Term, m_Br(m_ExtractValue<0>(m_CombineOr(
+                             m_Intrinsic<Intrinsic::amdgcn_if>(),
+                             m_Intrinsic<Intrinsic::amdgcn_else>())),
+                         m_Value(), m_Value()))) {
```
I'd rather do this in terms of the underlying IR, not the intrinsics we insert in control flow lowering. In the ideal future we will remove these and handle all control flow lowering in gMIR

ritter-x2a replied:
But that would mean reimplementing what SIAnnotateControlFlow does to determine if the intrinsics should be placed, right? I also considered integrating the functionality of this pass into SIAnnotateControlFlow, where we already know if we need to consider a branch, but I think that would add too much complexity to that pass.

ritter-x2a replied:
...I guess we could just consider all non-uniform (according to the Uniformity Analysis) branches.

arsenm replied:
I was guessing this would go in SIAnnotateControlFlow. But the check for the control flow intrinsics here seems redundant with isLikelyVarying below?

ritter-x2a commented on llvm/lib/Target/AMDGPU/AMDGPUAnnotateVaryingBranchWeights.cpp:
```diff
@@ -0,0 +1,201 @@
+  // Does it use a likely varying Value?
+  bool Result = false;
+  for (const auto &Use : U->operands()) {
+    Result |= isLikelyVarying(Use);
```
Right, good catch, thanks!

ritter-x2a replied:
With the updated heuristic in 8cf0cbf8fb44c9630cb48f51f3b2e22b68598272, recursion does not happen for phis any more, that should rule out endless loops.

jayfoad commented:
I still don't really understand...

Are you trying to identify branches whose condition is divergent according to UA, but is likely to be uniform in practice?

How can branch weights tell you whether the branch is likely uniform or not? I guess if it's likely uniform-and-true or uniform-and-false then they can tell you that.

For what kind of branches do you want to influence the exec-skipping logic? Can you give an example?

arsenm commented:
> Are you trying to identify branches whose condition is divergent according to UA, but is likely to be uniform in practice?

I think it's the opposite. It's divergent, and known never uniform in practice

ritter-x2a commented:
> I still don't really understand...
> 
> Are you trying to identify branches whose condition is divergent according to UA, but is likely to be uniform in practice?
> 
> How can branch weights tell you whether the branch is likely uniform or not? I guess if it's likely uniform-and-true or uniform-and-false then they can tell you that.
> 
> For what kind of branches do you want to influence the exec-skipping logic? Can you give an example?

There is an example in the added llvm/test/CodeGen/AMDGPU/conditional-mem-no-cbranch-execz.ll test (see also SWDEV-483228 for more discussion). When every wavefront diverges when executing a conditional, the execz branches to skip then/else cases are unnecessary and harm performance. The goal here is to remove the execz branches when that is likely the case (since the performance gains are probably not worth the effort of implementing a full, sound "must-not-be-uniform analysis" for this.)

ritter-x2a commented:
> How can branch weights tell you whether the branch is likely uniform or not? I guess if it's likely uniform-and-true or uniform-and-false then they can tell you that.

This patch doesn't use branch weights to find something out, it sets them to signal (to the later SIPreEmitPeephole pass) that it's likely that 'then' and 'else' case are executed, i.e., that it should remove the execz branches if possible.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAnnotateVaryingBranchWeights.cpp:
```diff
@@ -0,0 +1,201 @@
+
+private:
+  const GCNSubtarget &ST;
+  const TargetTransformInfo &TTI;
```
GCNTTI?

ritter-x2a replied:
Made the field obsolete and removed it in 8cf0cbf8fb44c9630cb48f51f3b2e22b68598272.

arsenm commented on llvm/test/CodeGen/AMDGPU/conditional-mem-no-cbranch-execz.ll:
```diff
@@ -0,0 +1,66 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx942 -mattr=+sramecc,-xnack < %s | FileCheck %s
```
Shouldn't need the target features

ritter-x2a replied:
Removed them in 345269b1523b1c2af48dc3798519031bf30d218d.

arsenm commented on llvm/test/CodeGen/AMDGPU/conditional-mem-no-cbranch-execz.ll:
```diff
@@ -0,0 +1,66 @@
+  ret void
+}
+
+attributes #0 = {"uniform-work-group-size"="true" "amdgpu-flat-work-group-size"="256,256"}
```
These shouldn't be necessary?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAnnotateVaryingBranchWeights.cpp:
```diff
@@ -0,0 +1,201 @@
+bool AMDGPUAnnotateVaryingBranchWeightsImpl::run(Function &F) {
+  // If the workgroup has only a single thread, the condition cannot vary.
+  const auto WGSizes = ST.getFlatWorkGroupSizes(F);
+  if (WGSizes.first <= 1)
```
Add test for this?

krzysz00 replied:
I also suspect you want the product here, not just checking the first term

ritter-x2a replied:
@krzysz00 From what I understand, `ST.getFlatWorkGroupSizes()` (based on the `amdgpu-flat-work-group-size` attribute) returns a pair of minimum and maximum numbers of work items (flattened across the dimensions), so `.first` is already the minimum number of total work items in the workgroup.
I'm currently working on a patch that refines the heuristic and also adds a similar check here based on the product of the `reqd_work_group_size` dimensions.

jayfoad replied:
Can you use `ST.isSingleLaneExecution` instead?

krzysz00 replied:
Ah, I misremembered which function you were calling

ritter-x2a replied:
> Can you use `ST.isSingleLaneExecution` instead?

Returning early if `ST.isSingleLaneExecution(F)` is `false` would mean annotating more cases: The intention of the current version was to annotate if it's known that single lane execution cannot happen for `F`. The changed variant would additionally annotate cases where we don't know if single lane execution could happen.
I'd be open to that change; it seems likely that kernels without explicit work group bounds and for which performance matters use more than a single lane.

arsenm commented on llvm/test/CodeGen/AMDGPU/conditional-mem-no-cbranch-execz.ll:
```diff
@@ -0,0 +1,66 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx942 -mattr=+sramecc,-xnack < %s | FileCheck %s
+
```
Need some IR pass checks

ritter-x2a replied:
Added tests for the pass in [annotate-likely-varying-branches.ll](https://github.com/llvm/llvm-project/pull/117567/commits/8cf0cbf8fb44c9630cb48f51f3b2e22b68598272#diff-561a228c0e2cb1220a56bc2619f9e0b87407a101372049e9ba68858381255183). I'll update them when changing to a different annotation mechanism instead of branch weights.

arsenm reviewed (commented):
I feel like this deserves separate representation from branch frequencies, but I'm not sure what it would be. And it would involve more threading work to get to the late backend pass

jayfoad commented:
I see. I think you really want to know "will this divergent condition be true for at least one lane", in which case the "execz" condition will not be satisfied so you can remove the s_cbranch_execz.

I agree that using normal branch weights for this feels wrong, since a condition like "laneid == 0" is mostly false (for most lanes) but it _does_ satisfy the condition you're interested in.

I don't see how any heuristic based on sources of divergence can help you here. Source of divergence just mean "not guaranteed to be uniform", and says nothing about whether the results in different lanes are actually likely to be different. I think perhaps you just need to pattern match specific cases like "laneid == constant" and "x == readfirstlane(x)".

jayfoad commented:
Do you have realistic examples of cases where we do want the execz branch, and cases where we do not want the execz branch?

I am sceptical that you can effectively distinguish these categories with a generic analysis of the IR (but maybe I'm wrong). I think it's more likely that you would have to pattern-match some examples of one category, and assume that everything else is in the other category.

ritter-x2a commented:
@jayfoad 
An example for where the execz branch **should** be eliminated would be a store in every even thread:

```cuda
__global__ void foo(int *buf, size_t n) {
    unsigned tid = blockDim.x * blockIdx.x + threadIdx.x;
    if (tid % 2 == 0)
        buf[tid] = 0;
}
```

An example where the execz branch **should not** be eliminated would be a sequence of conditional accesses where the condition is loaded from memory (the current implementation does not respect that):

```cuda
__global__ void bar(int *buf, size_t n, bool *lookup) {
    unsigned tid = blockDim.x * blockIdx.x + threadIdx.x;
    if (lookup[tid]) {
        buf[tid] = 0;
        // more memory accesses...
    }
}
```

Checking if the branch condition is the result of bitwise and arithmetic operations that involve a workitem id (i.e. only continuing to check the operands of a User if it is such an instruction) would be a plausible heuristic for that (that would benefit from benchmarking).

@krzysz00 might have more practical examples.

ritter-x2a commented:
> Not so sure about this. The compiler has no information about the contents of `lookup[tid]`, so no idea how likely it is that the condition will be false for every lane, correct?

Agreed. That's why I suggest keeping the handling for this case as it is in trunk and not eliminate the execz branch for it.

ritter-x2a committed:
fixup! [AMDGPU] Eliminate likely-spurious execz checks

First round of reviewer feedback: do not match intrinsics to identify candidate branches, preserve alphabetic order of the pass registry, rename TTI->GCNTTI, update test

ritter-x2a committed:
fixup! fixup! [AMDGPU] Eliminate likely-spurious execz checks

Make the heuristic more strict, add IR tests

ritter-x2a commented:
@jayfoad @arsenm I made the heuristic more selective in 8cf0cbf8fb44c9630cb48f51f3b2e22b68598272, it can now distinguish the example cases I mentioned above.

I haven't found a good alternative for using the branch weights yet since AFAIK normal IR metadata does not find its way to the backend. We could add a boolean `isLikelyVarying` parameter to the `amdgcn_if` and `amdgcn_else` intrinsics (I think that, in that case, moving the functionality from this pass into SIAnnotateControlFlow would make more sense).

krzysz00 commented on llvm/lib/Target/AMDGPU/AMDGPUAnnotateVaryingBranchWeights.cpp:
```diff
@@ -0,0 +1,257 @@
+//===----------------------------------------------------------------------===//
+// Estimate if conditional branches for which SIAnnotateControlFlow introduced
+// amdgcn_if or amdgcn_else intrinsics are likely to have different outcomes for
+// the threads of each wavefront. If that is the case, BranchWeight metadata is
```
Do we not usually call them lanes?

ritter-x2a replied:
Adjusted in 086ac46bce78a849eae39275c7601ad645783024.

krzysz00 reviewed (commented):
This feels like what I had in mind for the heuristic, but it seems like there's uncertainty there

ritter-x2a commented:
> This feels like what I had in mind for the heuristic, but it seems like there's uncertainty there

Thanks for the confirmation! I'm currently still working on getting performance numbers for a variant of this heuristic; I think the further steps on this PR depend on those performance numbers.

ritter-x2a committed:
Make comments refer to lanes instead of threads where lanes are meant.

ritter-x2a commented:
Closing this one in favor of #123749.
